estimagic/optimization/process_arguments.py

```diff
@@ -0,0 +1,311 @@
+
+        general_options (dict):
+            additional configurations for the optimization
+
+        algo_options (dict or list of dicts):
+            algorithm specific configurations for the optimization
+
+        dashboard (bool):
+            whether to create and show a dashboard
+
+        db_options (dict):
+            dictionary with kwargs to be supplied to the run_server function.
+    """
+
+    # set default arguments
+    criterion_kwargs = {} if criterion_kwargs is None else criterion_kwargs
+    constraints = [] if constraints is None else constraints
+    algo_options = {} if algo_options is None else algo_options
+    db_options = {} if db_options is None else db_options
+    general_options = {} if general_options is None else general_options
+
+    # Determine number of optimizations and check types
+
+    # Specify name and expected type for all arguments.
+    # Three groups of arguments are relevant:
+    # (i) expected as scalar,
+    # (ii) expected as scalar or list/tuple,
+    # (iii) expected as list/tuple or nested list/tuple
+
+    args_non_list = [
+        {"candidate": general_options, "scalar_type": dict, "name": "general_options"},
+        {"candidate": dashboard, "scalar_type": bool, "name": "dashboard"},
+        {"candidate": db_options, "scalar_type": dict, "name": "db_options"},
+    ]
+    args_list = [
+        {
+            "candidate": criterion,
+            "scalar_type": Callable,
+            "argument_required": True,
+            "name": "criterion",
+        },
+        {
+            "candidate": params,
+            "scalar_type": pd.DataFrame,
+            "argument_required": True,
+            "name": "params",
+        },
+        {
+            "candidate": algorithm,
+            "scalar_type": str,
+            "argument_required": True,
+            "name": "algorithm",
+        },
+        {
+            "candidate": algo_options,
+            "scalar_type": dict,
+            "argument_required": False,
+            "name": "algo_options",
+        },
+        {
+            "candidate": criterion_kwargs,
+            "scalar_type": dict,
+            "argument_required": False,
+            "name": "criterion_kwargs",
+        },
+    ]
+    args_nested_list = [
+        {"candidate": constraints, "argument_required": False, "name": "constraints"}
+    ]
+
+    # Check type of all inputs
+    for arg in args_non_list:
+        _check_type_nonlist_argument(**arg)
+
+    n_opts_args_list = [
+        _get_n_opt_and_check_type_list_argument(**arg) for arg in args_list
+    ]
+    n_opts_args_nested_list = [
+        _get_n_opt_and_check_type_nested_list_argument(**arg)
+        for arg in args_nested_list
+    ]
+
+    # Calc number of optimizations that should be run
+    n_opts = max(n_opts_args_list + n_opts_args_nested_list)
+
+    # Broadcast inputs
+    criterion, params, algorithm, algo_options, criterion_kwargs, constraints = [
```

Comment:
This is a matter of taste, but I prefer to use simple for loops and `append` instead of a list comprehension if the list comprehension spans more than one or two lines.